asmcomp/cmm_helpers.ml

```diff
@@ -1126,24 +1133,29 @@ let unaligned_set_16 ptr idx newval dbg =
     let v2 = Cop(Cand, [newval; cconst_int 0xFF], dbg) in
     let b1, b2 = if Arch.big_endian then v1, v2 else v2, v1 in
     Csequence(
-        Cop(Cstore (Byte_unsigned, Assignment), [add_int ptr idx dbg; b1], dbg),
+        Cop(Cstore (Byte_unsigned, Assignment), [add_int_addr ptr idx dbg; b1],
+            dbg),
         Cop(Cstore (Byte_unsigned, Assignment),
-            [add_int (add_int ptr idx dbg) (cconst_int 1) dbg; b2], dbg))
+            [add_int_addr (add_int_addr ptr idx dbg) (cconst_int 1) dbg; b2],
+            dbg))
 
 let unaligned_load_32 ptr idx dbg =
   if Arch.allow_unaligned_access
-  then Cop(mk_load_mut Thirtytwo_unsigned, [add_int ptr idx dbg], dbg)
+  then Cop(mk_load_mut Thirtytwo_unsigned, [add_int_addr ptr idx dbg], dbg)
   else
     let cconst_int i = Cconst_int (i, dbg) in
-    let v1 = Cop(mk_load_mut Byte_unsigned, [add_int ptr idx dbg], dbg) in
+    let v1 = Cop(mk_load_mut Byte_unsigned, [add_int_addr ptr idx dbg], dbg) in
     let v2 = Cop(mk_load_mut Byte_unsigned,
-                 [add_int (add_int ptr idx dbg) (cconst_int 1) dbg], dbg)
+                 [add_int_addr (add_int_addr ptr idx dbg) (cconst_int 1) dbg],
+                 dbg)
```

Comment:
@gasche was wondering whether to associate differently, i.e. `ptr + (idx + 1)` instead of the current `(ptr + idx) + 1`, so that `(idx + 1)` is simplified if `idx` is a constant.   But if `idx` is not a constant, we would lose the common subexpression `(ptr + idx)`, which CSE can factor.  So, my take is that this code should not be changed, but `add_int_addr` should treat the case `(ptr + cst) + cst` specially.